api/src/OpenTelemetry/Internal/Logs/Types.hs

```diff
@@ -19,20 +23,114 @@ module OpenTelemetry.Internal.Logs.Types (
+  -- mean that the task of exporting can not be done concurrently. How this is done is outside the scope of this specification.
+  -- Each implementation MUST document the concurrency characteristics the SDK requires of the exporter.
+  --
+  -- Export MUST NOT block indefinitely, there MUST be a reasonable upper limit after which the call must time out with an error result (Failure).
+  --
+  -- Concurrent requests and retry logic is the responsibility of the exporter. The default SDK’s LogRecordProcessors SHOULD NOT implement
+  -- retry logic, as the required logic is likely to depend heavily on the specific protocol and backend the logs are being sent to.
+  -- For example, the OpenTelemetry Protocol (OTLP) specification defines logic for both sending concurrent requests and retrying requests.
+  --
+  -- Result:
+  -- Success - The batch has been successfully exported. For protocol exporters this typically means that the data is sent over the wire and delivered to the destination server.
+  -- Failure - exporting failed. The batch must be dropped. For example, this can happen when the batch contains bad data and cannot be serialized.
+  , logRecordExporterForceFlushInternal :: IO ()
+  -- ^ This is a hint to ensure that the export of any ReadableLogRecords the exporter has received prior to the call to ForceFlush SHOULD
+  -- be completed as soon as possible, preferably before returning from this method.
+  --
+  -- ForceFlush SHOULD provide a way to let the caller know whether it succeeded, failed or timed out.
+  --
+  -- ForceFlush SHOULD only be called in cases where it is absolutely necessary, such as when using some FaaS providers that may suspend
+  -- the process after an invocation, but before the exporter exports the ReadlableLogRecords.
+  --
+  -- ForceFlush SHOULD complete or abort within some timeout. ForceFlush can be implemented as a blocking API or an asynchronous API which
+  -- notifies the caller via a callback or an event. OpenTelemetry SDK authors MAY decide if they want to make the flush timeout configurable.
+  , logRecordExporterShutdownInternal :: IO ()
+  -- ^
+  -- Shuts down the exporter. Called when SDK is shut down. This is an opportunity for exporter to do any cleanup required.
+  --
+  -- Shutdown SHOULD be called only once for each LogRecordExporter instance. After the call to Shutdown subsequent calls to Export are not
+  -- allowed and SHOULD return a Failure result.
+  --
+  -- Shutdown SHOULD NOT block indefinitely (e.g. if it attempts to flush the data and the destination is unavailable).
+  -- OpenTelemetry SDK authors MAY decide if they want to make the shutdown timeout configurable.
+  }
+
+
+-- | @LogRecordExporter@s are an @MVar@ wrapper around @LogRecordExporterInternal@ that prevent @logRecordExporterExport@ from being called concurrently for the same exporter.
+newtype LogRecordExporter = LogRecordExporter {unExporter :: MVar LogRecordExporterInternal}
```

Comment:
This does not seem super clean, so maybe it could be done in a different way. I don't think it would be possible to write `withLogRecordExporter` as a similar function to `withMVar` because then a developer could call `logRecordExporterExport` concurrently within `withLogRecordExporter`.